corehq/motech/repeaters/tasks.py

```diff
@@ -322,7 +323,99 @@ def get_repeater_lock(repeater_id):
+    repeat_records = repeater.repeat_records_ready[:repeater.num_workers]
+    header_tasks = [get_task_signature(rr) for rr in repeat_records]
+    chord(header_tasks)(update_repeater.s(repeater_id, lock_token))
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_pending_repeat_record(repeat_record_id, domain):
+    # NOTE: Keep separate from `process_failed_repeat_record()` for
+    # monitoring purposes. `domain` is for tagging in Datadog
+    return process_ready_repeat_record(repeat_record_id)
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_failed_repeat_record(repeat_record_id, domain):
+    # NOTE: Keep separate from `process_pending_repeat_record()` for
+    # monitoring purposes. `domain` is for tagging in Datadog
+    return process_ready_repeat_record(repeat_record_id)
+
+
+def process_ready_repeat_record(repeat_record_id):
+    state_or_none = None
+    with TimingContext('process_repeat_record') as timer:
+        try:
+            repeat_record = (
+                RepeatRecord.objects
+                .prefetch_related('repeater', 'attempt_set')
+                .get(id=repeat_record_id)
+            )
+            report_repeater_attempt(repeat_record.repeater.repeater_id)
+            if not is_repeat_record_ready(repeat_record):
+                return None
+            with timer('fire_timing') as fire_timer:
+                state_or_none = repeat_record.fire(timing_context=fire_timer)
+            report_repeater_usage(
+                repeat_record.domain,
+                # round up to the nearest millisecond, meaning always at least 1ms
+                milliseconds=int(fire_timer.duration * 1000) + 1
+            )
+        except Exception:
+            logging.exception(f'Failed to process repeat record {repeat_record_id}')
+    return state_or_none
+
+
+def is_repeat_record_ready(repeat_record):
+    # Fail loudly if repeat_record is not ready.
+    # process_ready_repeat_record() will log an exception.
+    assert repeat_record.state in (State.Pending, State.Fail)
+
+    # The repeater could have been paused or rate-limited while it was
+    # being processed
+    return (
+        not repeat_record.repeater.is_paused
+        and not toggles.PAUSE_DATA_FORWARDING.enabled(repeat_record.domain)
+        and not rate_limit_repeater(
+            repeat_record.domain,
+            repeat_record.repeater.repeater_id
+        )
+    )
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def update_repeater(repeat_record_states, repeater_id, lock_token):
+    """
+    Determines whether the repeater should back off, based on the
+    results of ``_process_repeat_record()`` tasks.
+    """
+    try:
+        repeater = Repeater.objects.get(id=repeater_id)
+        if any(s == State.Success for s in repeat_record_states):
+            # At least one repeat record was sent successfully. The
+            # remote endpoint is healthy.
+            repeater.reset_backoff()
```

Comment:
I don't think so. If an API always responds slowly then we'll end up only sending `repeater.num_workers` repeat records [once every seven days](https://github.com/dimagi/commcare-hq/blob/0db6a6c4f4ca90d3650f7d7beaf753042b97df8f/corehq/motech/repeaters/const.py#L7) (the maximum back-off interval). That seems wrong.

Compare that to the current behavior, where we will always try to send when we register a repeat record, even if every single repeat record has failed before.

Backing off from a healthy endpoint is a scenario that really worries me about this new process, because it could mean that no payloads go through for long intervals, and that can have strong negative consequences.